presto-main/src/main/java/com/facebook/presto/operator/aggregation/OptimizedTypedSet.java

```diff
@@ -0,0 +1,355 @@
+            if (hashPosition != INVALID_POSITION) {
+                // There is no need to test if adding element is successful since it's on the same hash table
+                addElement(blockPositionByHash, hashPosition, block, i);
+                positions[positionsIndex++] = i;
+            }
+        }
+
+        getPositionsForBlocks().add(positionsList(positions, 0, positionsIndex));
+        size += positionsIndex;
+    }
+
+    /**
+     * Intersect the set with the elements of the block, eliminating duplicates.
+     */
+    public void intersect(Block block)
+    {
+        int positionCount = block.getPositionCount();
+
+        currentBlockIndex++;
+        ensureBlocksCapacity(currentBlockIndex + 1);
+        blocks[currentBlockIndex] = block;
+
+        if (currentBlockIndex == 0) {
+            // This set was an empty set, so the result set should also be an empty set.
+            positionsForBlocks.add(EMPTY_SELECTED_POSITIONS);
+            return;
+        }
+        else {
+            int[] positions = ensureCapacity(positionsForBlocks.get(currentBlockIndex - 1).getPositions(), positionCount);
+
+            // We need to create a new hash table because the elements in the set may be removed
+            long[] newBlockPositionByHash = initializeHashTable();
+
+            int positionsIndex = 0;
+            for (int i = 0; i < positionCount; i++) {
+                int hash = getMaskedHash(hashPosition(elementType, block, i));
+                int positionInBlockPositionByHash = getInsertPosition(blockPositionByHash, hash, block, i);
+                if (positionInBlockPositionByHash == INVALID_POSITION) {
+                    // add to the hash table if it exists in blockPositionByHash
+                    if (addElement(newBlockPositionByHash, hash, block, i)) {
+                        positions[positionsIndex++] = i;
+                    }
+                }
+            }
+
+            blockPositionByHash = newBlockPositionByHash;
+            getPositionsForBlocks().add(positionsList(positions, 0, positionsIndex));
+            size = positionsIndex;
+
+            clearPreviousBlocks();
+        }
+    }
+
+    private long[] initializeHashTable()
+    {
+        long[] newBlockPositionByHash = new long[hashCapacity]; // Create a new hashtable
+        Arrays.fill(newBlockPositionByHash, EMPTY_SLOT);
+        return newBlockPositionByHash;
+    }
+
+    /**
+     * Add the elements of the block that do not exist in the set, eliminating duplicates, and remove all previously existing elements.
+     */
+    public void except(Block block)
+    {
+        int positionCount = block.getPositionCount();
+
+        if (currentBlockIndex == -1) {
+            // This set was an empty set. Call union() to remove duplicates.
+            union(block);
+            return;
+        }
+        else {
```

Comment:
There's common functionality between ```except``` and ```intersect``` eg: the ```for``` loop. Can we try to extract the common things to functions - makes it more readable and reduces changes of bugs :)

Reply:
@sujay-jain I understand they look similar, but the logic is different. `intersect` has
```
if (positionInBlockPositionByHash == INVALID_POSITION) {
   // Add the new element
}
```
But the `except` has 
```
if (positionInBlockPositionByHash != INVALID_POSITION) {
   // Add the new element
}
```
This makes it hard to extract common method. The only part can be extracted was 
```
if (addElement(hashtable, hash, block, position)) {
        positions[positionsIndex++] = position;
}
```
I tried to extract it to a private method: 
```
private int addElementIfNotExists(Block block, int position, int[] positions, long[] hashtable, int hash, int positionsIndex)
    {
        if (addElement(hashtable, hash, block, position)) {
            positions[positionsIndex++] = position;
        }
        return positionsIndex;
    }
```
And now the callsites become
intersect:
```
for (int i = 0; i < positionCount; i++) {
                int hash = getMaskedHash(hashPosition(elementType, block, i));
                int positionInBlockPositionByHash = getInsertPosition(blockPositionByHash, hash, block, i);
                if (positionInBlockPositionByHash == INVALID_POSITION) {
                    // add to the hash table if it exists in blockPositionByHash
                    positionsIndex = addElementIfNotExists(block, i, positions, newBlockPositionByHash, hash, positionsIndex);
                }
            }
```
exept:
```
for (int i = 0; i < positionCount; i++) {
                int hash = getMaskedHash(hashPosition(elementType, block, i));
                int positionInBlockPositionByHash = getInsertPosition(blockPositionByHash, hash, block, i);

                // add to the hash table if it does not exist in blockPositionByHash
                if (positionInBlockPositionByHash != INVALID_POSITION) {
                    positionsIndex = addElementIfNotExists(block, i, positions, newBlockPositionByHash, hash, positionsIndex);
                }
            }
```
Now you have to jump and scroll down to check what `addElementIfNotExists` does. Its 6 parameters makes it hard to know what's done inside. Also, it mixes operations on the hash table with the operation on `positions` array which are two different objects in the class that serves different purposes. It doesn't save lines either: the two lines that got extracted into a function just adds inconvenience for the readers. 

Combining these considerations I think keeping it as original may be better. If you see other ways doing this, please let me know!

Reply:
ok, I think we can leave it as is if you feel that's cleaner. Thanks for trying it.